### DISCUSSION TITLE
How to store the AST?

### DISCUSSION THREAD

**Discussion by mortenson** [google/cel-go#897] [Q&A] [answered]
Hi there, I'm just starting to use cel-go and was reading https://codelabs.developers.google.com/codelabs/cel-go#1, which mentions

>There are three phases in processing an expression: parse, check, and evaluate. The most common pattern for CEL is for a control plane to parse and check expressions at config time, and store the AST.

How should the AST be stored? Is there a code example for serializing the AST?

**Comment by jnthntatum** [comment, marked as answer]
It's normally stored or transferred between programs in the protobuf format. See [cel.AstToCheckedExpr](https://github.com/google/cel-go/blob/master/cel/io.go#L57) for converting from the go object to proto. The conformance server harness has some examples (though the proto serialization is handled by grpc). See https://github.com/google/cel-go/blob/master/server/server.go and https://github.com/google/cel-spec/blob/master/tools/celrpc/celrpc.go

The [protobuf dev guide](https://protobuf.dev/getting-started/gotutorial/#writing-a-message) has examples for reading/writing the serialized messages.

**Reply by mortenson** [reply]
Thank you!
